website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+
+- Input: _"List all customers in New York."_
+
+- Generated SQL: _"SELECT name FROM customers WHERE location = 'New York';"_
+
+**Code:**
+
+```python
+def generate(state: GraphState):
+    print("---GENERATING SQL QUERY---")
+    messages = state["messages"]
+    iterations = state["iterations"]
+    translated_input = state["translated_input"]
+
+    # Retrieve relevant documents from the vector store using the translated input
+    docs = vector_store.similarity_search(translated_input, k=4)
+    retrieved_docs = "\n\n".join([doc.page_content for doc in docs])
+
+    # Generate the SQL query using the chain
+    sql_solution = sql_gen_chain.invoke(
+        {
+            "retrieved_docs": retrieved_docs,
+            "messages": [("user", translated_input)],
+        }
+    )
+
+    # Save the generated SQL query in the state
+    messages += [
+        (
+            "assistant",
+            f"{sql_solution.description}\nSQL Query:\n{sql_solution.sql_code}",
+        )
+    ]
+    iterations += 1
+
+    state["generation"] = sql_solution
+    state["messages"] = messages
+    state["iterations"] = iterations
+
+    return state
+```
+
+#### **6. SQL Check**
+
+The `sql_check` node validates the generated SQL query for safety and integrity before execution.
+
+**Purpose:** Ensure the SQL query adheres to safety and syntactical standards.
+
+**Process:**
+
+- Executes the query within a transactional savepoint to test its validity.
+
+- Rolls back any changes after validation.
+
+- Flags errors and updates the state if validation fails.
+
+**Key Considerations:**
+
+- Detects potentially destructive operations.
+
+- Provides detailed feedback on validation errors.
+
+**Examples:**
+
+- Input SQL: _"SELECT name FROM customers WHERE city = 'New York';"_
+
+- Validation: Query is valid.
+
+- Input SQL: _"SELECT MONTH(date) AS month, SUM(total) AS total_sales FROM orders GROUP BY MONTH(date);"_
+
+- Response: _"Your SQL query failed to execute: no such function: MONTH."_
+
+**Code:**
+
+```python
+def sql_check(state: GraphState):
+    print("---VALIDATING SQL QUERY---")
+    messages = state["messages"]
+    sql_solution = state["generation"]
+    error = "no"
+
+    sql_code = sql_solution.sql_code.strip()
+
+    try:
+        # Start a savepoint for the transaction
+        conn.execute('SAVEPOINT sql_check;')
+        # Attempt to execute the SQL query
+        cursor.execute(sql_code)
+        # Roll back to the savepoint to undo any changes
+        conn.execute('ROLLBACK TO sql_check;')
+        print("---SQL QUERY VALIDATION: SUCCESS---")
+    except Exception as e:
+        # Roll back in case of error
+        conn.execute('ROLLBACK TO sql_check;')
+        print("---SQL QUERY VALIDATION: FAILED---")
+        print(f"Error: {e}")
+        messages += [("user", f"Your SQL query failed to execute: {e}")]
+        error = "yes"
+
+    state["error"] = error
+
+    return state
+```
```

Comment:
This seems to validate that the SQL runs, but I don't see where it checks for safety or detects potentially destructive operations. As far as I can tell, a drop table command would clear this step. See earlier note—it looks like the safety check occurs before the sql is generated.